README.rst
MS Outlook Web access client
============================

Synopsis
--------

Stuck with Microsoft Outlook for work e-mail? Want access to your raw e-mail messages from Outlook? Prefer to use your own e-mail client rather than the clumsy Outlook client? If so, weboutlook is for you.

`weboutlook` is a Python module that retrieves full, raw e-mails from Microsoft Outlook Web Access by screen scraping. It can do the following:

Log into a Microsoft Outlook Web Access account with a given username and password.
Retrieve all e-mail IDs from the first page of your Inbox.
Retrieve the full, raw source of the e-mail with a given ID.
Delete an e-mail with a given ID (technically, move it to the 'Deleted Items' folder).
Documentation / sample usage::

    >>> from weboutlook import scraper

    # Throws InvalidLogin exception for invalid username/password.
    >>> s = scraper('https://webmaildomain.com', 'username', 'invalid password')
    >>> s.login() Traceback (most recent call last):
    ...
    scraper.InvalidLogin

    >>> s = scraper('https://webmaildomain.com', 'username', 'correct password')
    >>> s.login()

    # Display IDs of messages in the inbox.
    >>> s.inbox()
    ['/Inbox/test-3.EML']

    # Display IDs of messages in the 'sent items' folder.
    >>> s.get_folder('sent items')
    '/Sent%20Items/test-2.EML'

    # Display the raw source of a particular message.
    >>> print s.get_message('/Inbox/Hey%20there.EML')
    ...

    # Delete a message.
    >>> s.delete_message('/Inbox/Hey%20there.EML')



weboutlook/changelog.txt
Changelog (0.33-SPH)
====================

Submitted by Sam Holland (samuar@gmail.com)

[popdaemon.py]
Implemented per-msg LIST commands as per RFC1939[1]
Implemented per-msg UIDL commands as per RFC1939[1]

[scraper.py]
Line 90 Changed to fix newline in Cookie output as per Issue 2 of [2] (for my domain mail.bader.mod.uk)


References
[1] (http://www.ietf.org/rfc/rfc1939.txt)
[2] (http://code.google.com/p/weboutlook/)


Changelog (0.4-SPH)
===================

Tested with Outlook 2007, Outlook Express 6, Thunderbird and Evolution

[popdaemon.py]
Implemented CAPA command [Lists USER and UIDL]
Fixed LIST and UIDL commands
	These returned a "." for a single line message, which tripped up Outlook Express 6.

weboutlook/popdaemon.py
"""
POP e-mail server for Microsoft Outlook Web Access scraper

This wraps the Outlook Web Access scraper by providing a POP interface to it.

Run this file from the command line to start the server on localhost port 8110.
The server will run until you end the process. If you specify the "--once"
command-line option, it'll only run one full POP transaction and quit. The
latter is useful for the "precommand" option in KMail, which you can configure
to run this script every time you check e-mail so it starts the server in the
background.

Note that you'll have to specify WEBMAIL_SERVER in this file.
"""

# Copyright (C) 2006 Adrian Holovaty <holovaty at gmail dot com>, and 2007 
# Cloudburst, LLC <evan at cloudbur dot st>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA 02111-1307 USA

import asyncore, asynchat, socket, sys, re, pdb

# relative import
from scraper import InvalidLogin, OutlookWebScraper

__version__ = 'samuar.com OWA-POP3 Server v0.4(Beta)'

TERMINATOR = '\r\n'
WEBMAIL_SERVER = 'https://mail.bader.mod.uk/'

def quote_dots(lines):
    for line in lines:
        if line.startswith("."):
            line = "." + line
        yield line

class POPChannel(asynchat.async_chat):
    def __init__(self, conn, quit_after_one):
        self.quit_after_one = quit_after_one
        asynchat.async_chat.__init__(self, conn)
        self.__line = []
        self.push('+OK %s %s' % (socket.getfqdn(), __version__))
        self.set_terminator(TERMINATOR)
        self._activeDataChannel = None

    # Overrides base class for convenience
    def push(self, msg):
#        print("said: %s" % msg)
        asynchat.async_chat.push(self, msg + TERMINATOR)

    # Implementation of base class abstract method
    def collect_incoming_data(self, data):
        self.__line.append(data)

    # Implementation of base class abstract method
    def found_terminator(self):
        line = ''.join(self.__line)
        self.__line = []
        if not line:
            self.push('500 Error: bad syntax')
            return
        method = None
        i = line.find(' ')
        if i < 0:
            command = line.upper()
            arg = None
        else:
            command = line[:i].upper()
            arg = line[i+1:].strip()
        print(' got: %s' % line)        
        method = getattr(self, 'pop_' + command, None)
        if not method:
            self.push('-ERR Error : command "%s" not implemented' % command)
            return
        method(arg)
        return

    def pop_USER(self, user):
        # Logs in any username.
        if not user:
            self.push('-ERR: Syntax: USER username')
        else:
            self.username = user # Store for later.
            self.push('+OK Password required')

    def pop_PASS(self, password=''):
        self.scraper = OutlookWebScraper(WEBMAIL_SERVER, self.username, password)
        try:
            self.scraper.login()
        except InvalidLogin:
            self.push('-ERR Login failed. (Wrong username/password?)')
        else:
            self.push('+OK User logged in')
            self.inbox_cache = self.scraper.inbox()
            self.msg_cache = [self.scraper.get_message(msg_id) for msg_id in self.inbox_cache]

    def pop_STAT(self, arg):
            dropbox_size = sum([len(msg) for msg in self.msg_cache])
            self.push('+OK %d %d' % (len(self.inbox_cache), dropbox_size))

    def pop_NOOP(self, arg):
        self.push('+OK')

    def pop_LIST(self, arg):
        if not arg:
            num_messages = len(self.inbox_cache)
            self.push('+OK')
            for i, msg in enumerate(self.msg_cache):
                self.push('%d %d' % (i+1, len(msg)))
            self.push(".")
        else:
            found = False
            for i, msg in enumerate(self.msg_cache):
                if (i+1) == int(arg):
                    self.push('+OK %d %d' % (i+1, len(msg)))
                    found = True
            if not found:
                self.push('-ERR no such message, only %d messages in maildrop' %(len(self.inbox_cache)))
            #self.push(".")


    def pop_UIDL(self, arg):
        if not arg:
            num_messages = len(self.inbox_cache)
            self.push('+OK')
            for i, msg in enumerate(self.msg_cache):
                self.push('%d %s' % (i+1, re.search(r"Message-ID: <(.*?)>", msg, re.S | re.I).group(1)))
            self.push(".")
        else:
            found = False
            for i, msg in enumerate(self.msg_cache):
                if (i+1) == int(arg):
                    self.push('+OK %d %s' % (i+1, re.search(r"Message-ID: <(.*?)>", msg, re.S | re.I).group(1)))
                    found = True
            if not found:
                self.push('-ERR no such message, only %d messages in maildrop' % (len(self.inbox_cache)))
            #self.push(".")

    def pop_RETR(self, arg):
        if not arg:
            self.push('-ERR: Syntax: RETR msg')
            print '-ERR: Syntax: RETR msg'
        else:
            # TODO: Check request is in range.
            msg_index = int(arg) - 1
            msg = self.msg_cache[msg_index]
            msg_id = self.inbox_cache[msg_index]
            msg = msg.lstrip() + TERMINATOR

            self.push('+OK')
            print '+OK (pop_RETR l_138)'

            for line in quote_dots(msg.split(TERMINATOR)):
                self.push(line)
            self.push('.')

    def pop_CAPA(self, arg):
        self.push('+OK Capability list follows')
        self.push('USER')
        self.push('UIDL')
        self.push('.')
            
    def pop_DELE(self, arg):
        if not arg:
            self.push('-ERR: Syntax: RETR msg')
        else:
            msg_index = int(arg) - 1
            msg_id = self.inbox_cache[msg_index]        
            self.scraper.delete_message(msg_id)
            self.push('+OK')
#            self.push(".")

    def pop_QUIT(self, arg):
        self.push('+OK Goodbye')
        self.close_when_done()
        if self.quit_after_one:
            # This SystemExit gets propogated to handle_error(),
            # which stops the program. Slightly hackish.
            raise SystemExit

    def handle_error(self):
        if self.quit_after_one:
            sys.exit(0) # Exit.
        else:
            asynchat.async_chat.handle_error(self)

class POP3Proxy(asyncore.dispatcher):
    def __init__(self, localaddr, quit_after_one):
        """
        localaddr is a tuple of (ip_address, port).

        quit_after_one is a boolean specifying whether the server should quit
        after serving one session.
        """
        self.quit_after_one = quit_after_one
        asyncore.dispatcher.__init__(self)
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        # try to re-use a server port if possible
        self.set_reuse_addr()
        self.bind(localaddr)
        self.listen(5)

    def handle_accept(self):
        conn, addr = self.accept()
        channel = POPChannel(conn, self.quit_after_one)

if __name__ == '__main__':
    from optparse import OptionParser
    parser = OptionParser("usage: %prog [options]")
    parser.add_option('--once', action='store_true', dest='once',
        help='Serve one POP transaction and then quit. (Server runs forever by default.)')
    options, args = parser.parse_args()
    proxy = POP3Proxy(('0.0.0.0', 2221), options.once is True)
    try:
        asyncore.loop()
    except KeyboardInterrupt:
        pass


weboutlook/scraper.py
"""
Microsoft Outlook Web Access scraper

Retrieves full, raw e-mails from Microsoft Outlook Web Access by
screen scraping. Can do the following:

    * Log into a Microsoft Outlook Web Access account with a given username
      and password.
    * Retrieve all e-mail IDs from the first page of your Inbox.
    * Retrieve the full, raw source of the e-mail with a given ID.
    * Delete an e-mail with a given ID (technically, move it to the "Deleted
      Items" folder).

The main class you use is OutlookWebScraper. See the docstrings in the code
and the "sample usage" section below.

This module does no caching. Each time you retrieve something, it does a fresh
HTTP request. It does cache your session, though, so that you only have to log
in once.
"""

# Documentation / sample usage:
#
# # Throws InvalidLogin exception for invalid username/password.
# >>> s = OutlookWebScraper('https://webmaildomain.com', 'username', 'invalid password')
# >>> s.login()
# Traceback (most recent call last):
#     ...
# scraper.InvalidLogin
#
# >>> s = OutlookWebScraper('https://webmaildomain.com', 'username', 'correct password')
# >>> s.login()
#
# # Display IDs of messages in the inbox.
# >>> s.inbox()
# ['/Inbox/Hey%20there.EML', '/Inbox/test-3.EML']
#
# # Display IDs of messages in the "sent items" folder.
# >>> s.get_folder('sent items')
# ['/Sent%20Items/test-2.EML']
#
# # Display the raw source of a particular message.
# >>> print s.get_message('/Inbox/Hey%20there.EML')
# [...]
#
# # Delete a message.
# >>> s.delete_message('/Inbox/Hey%20there.EML')

# Copyright (C) 2006 Adrian Holovaty <holovaty@gmail.com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA 02111-1307 USA

import re, socket, urllib, urlparse
from Cookie import SimpleCookie

__version__ = '0.1'
__author__ = 'Adrian Holovaty <holovaty@gmail.com>'

socket.setdefaulttimeout(15)

class InvalidLogin(Exception):
    pass

class RetrievalError(Exception):
    pass

class CookieScraper(object):
    "Scraper that keeps track of getting and setting cookies."
    def __init__(self):
        self._cookies = SimpleCookie()

    def get_page(self, url, post_data=None, headers=()):
        """
        Helper method that gets the given URL, handling the sending and storing
        of cookies. Returns the requested page as a string.
        """
        opener = urllib.URLopener()
        opener.addheader('Cookie', self._cookies.output(attrs=[], header='').strip().replace('\r\n', '; '))
        for k, v in headers:
            opener.addheader(k, v)
        try:
            f = opener.open(url, post_data)
        except IOError, e:
            if e[1] == 302:
                # Got a 302 redirect, but check for cookies before redirecting.
                # e[3] is a httplib.HTTPMessage instance.
                if e[3].dict.has_key('set-cookie'):
                    self._cookies.load(e[3].dict['set-cookie'])
                return self.get_page(e[3].getheader('location'))
            else:
                raise
        if f.headers.dict.has_key('set-cookie'):
            self._cookies.load(f.headers.dict['set-cookie'])
        return f.read()

class OutlookWebScraper(CookieScraper):
    def __init__(self, domain, username, password):
        self.domain = domain
        self.username, self.password = username, password
        self.is_logged_in = False
        self.base_href = None
        super(OutlookWebScraper, self).__init__()

    def login(self):
        url = urlparse.urljoin(self.domain, 'exchweb/bin/auth/owaauth.dll')
        html = self.get_page(url, urllib.urlencode({
            'destination': urlparse.urljoin(self.domain, 'exchange'),
            'flags': '0',
            'username': self.username,
            'password': self.password,
            'SubmitCreds': 'Log On',
            'forcedownlevel': '0',
            'trusted': '4',
        }))
        if 'You could not be logged on to Outlook Web Access' in html:
            raise InvalidLogin
        m = re.search(r'(?i)<BASE href="([^"]*)">', html)
        if not m:
            raise RetrievalError, "Couldn't find <base href> on page after logging in."
        self.base_href = m.group(1)
        self.is_logged_in = True

    def inbox(self):
        """
        Returns the message IDs for all messages on the first page of the
        Inbox, regardless of whether they've already been read.
        """
        return self.get_folder('Inbox')

    def get_folder(self, folder_name):
        """
        Returns the message IDs for all messages on the first page of the
        folder with the given name, regardless of whether the messages have
        already been read. The folder name is case insensitive.
        """
        if not self.is_logged_in: self.login()
        url = self.base_href + urllib.quote(folder_name) + '/?Cmd=contents'
        html = self.get_page(url)
        message_urls = re.findall(r'(?i)NAME=MsgID value="([^"]*)"', html)
        return message_urls

    def get_message(self, msgid):
        "Returns the raw e-mail for the given message ID."
        if not self.is_logged_in: self.login()
        # Sending the "Translate=f" HTTP header tells Outlook to include
        # full e-mail headers. Figuring that out took way too long.
        return self.get_page(self.base_href + msgid + '?Cmd=body', headers=[('Translate', 'f')])

    def delete_message(self, msgid):
        "Deletes the e-mail with the given message ID."
        if not self.is_logged_in: self.login()
        return self.get_page(self.base_href + msgid, urllib.urlencode({
            'MsgId': msgid,
            'Cmd': 'delete',
            'ReadForm': '1',
        }))
